test/middleware.js

```diff
@@ -417,4 +420,255 @@ describe('middleware', function() {
       doc.submitOp([{p: ['tricks', 1], li: 'stay'}], {source: 'c'}, errorHandler(done));
     });
   });
+
+  describe('$fixup', function() {
+    var connection;
+    var backend;
+    var doc;
+
+    beforeEach(function(done) {
+      backend = this.backend;
+      connection = backend.connect();
+      doc = connection.get('dogs', 'fido');
+
+      doc.create({name: 'fido'}, done);
+    });
+
+    it('applies a fixup op to the client that submitted it', function(done) {
+      backend.use('apply', function(request, next) {
+        next(request.$fixup([{p: ['tricks', 1], li: 'stay'}]));
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], function(error) {
+        if (error) return done(error);
+        expect(doc.data.tricks).to.eql(['fetch', 'stay']);
+        expect(doc.version).to.equal(2);
+        done();
+      });
+    });
+
+    it('emits an op event for the fixup', function(done) {
+      backend.use('apply', function(request, next) {
+        next(request.$fixup([{p: ['tricks', 1], li: 'stay'}]));
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], errorHandler(done));
+
+      doc.on('op', function() {
+        expect(doc.data.tricks).to.eql(['fetch', 'stay']);
+        done();
+      });
+    });
+
+    it('passes the fixed up op to future middleware', function(done) {
+      backend.use('apply', function(request, next) {
+        next(request.$fixup([{p: ['tricks', 1], li: 'stay'}]));
+      });
+
+      backend.use('apply', function(request, next) {
+        expect(request.op.op).to.eql([
+          {p: ['tricks'], oi: ['fetch']},
+          {p: ['tricks', 1], li: 'stay'}
+        ]);
+        next();
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], done);
+    });
+
+    it('applies the composed op to a remote client', function(done) {
+      backend.use('apply', function(request, next) {
+        next(request.$fixup([{p: ['tricks', 1], li: 'stay'}]));
+      });
+
+      var remoteConnection = backend.connect();
+      var remoteDoc = remoteConnection.get('dogs', 'fido');
+
+      remoteDoc.subscribe(function(error) {
+        if (error) return done(error);
+
+        expect(remoteDoc.data).to.eql({name: 'fido'});
+
+        remoteDoc.on('op batch', function() {
+          expect(remoteDoc.data.tricks).to.eql(['fetch', 'stay']);
+          expect(doc.version).to.equal(remoteDoc.version);
+          done();
+        });
+
+        doc.submitOp([{p: ['tricks'], oi: ['fetch']}], errorHandler(done));
+      });
+    });
+
+    it('transforms pending ops by the fixup', function(done) {
```

Comment:
We should also add a test case that locally we transform correctly. The following case requires us to reorder our apply and transform:

```js
    it('transforms pending ops by the fixup', function(done) {
      var applied = false;
      backend.use('apply', function(request, next) {
        if (applied) return next();
        applied = true;
        next(request.$fixup([{p: ['tricks', 1], li: 'stay'}]));
      });

      var remoteConnection = backend.connect();
      var remoteDoc = remoteConnection.get('dogs', 'fido');

      remoteDoc.subscribe(function(error) {
        if (error) return done(error);

        expect(remoteDoc.data).to.eql({name: 'fido'});

        remoteDoc.on('op batch', function() {
          if (remoteDoc.version !== 3) return;
          expect(remoteDoc.data.tricks).to.eql(['stay']);
          expect(remoteDoc.data).to.eql(doc.data);
          done();
        });

        doc.preventCompose = true;
        doc.submitOp([{p: ['tricks'], oi: ['fetch']}], errorHandler(done));
        doc.submitOp([{p: ['tricks', 0], ld: 'fetch'}], errorHandler(done));
      });
    });
```

Reply:
If I add logs here:

```js
  if (message.fu) {
    console.log(this.data);
    console.log(message.fu[0].op);
    for (var i = 0; i < message.fu.length; i++) {
      var fixupOp = message.fu[i];
      try {
        this._otApply(fixupOp, false);
      } catch (error) {
        return this._hardRollback(error);
      }

      for (var j = 0; j < this.pendingOps.length; j++) {
        var transformErr = transformX(this.pendingOps[i], fixupOp);
        if (transformErr) return this._hardRollback(transformErr);
      }
    }
  }
```

This is what I get:

```js
{ name: 'fido', tricks: [] }
[ { p: [ 'tricks', 1 ], li: 'stay' } ]
```

Reply:
If I then update to:

```js
  if (message.fu) {
    for (var i = 0; i < message.fu.length; i++) {
      var fixupOp = message.fu[i];
      for (var j = 0; j < this.pendingOps.length; j++) {
        var transformErr = transformX(this.pendingOps[i], fixupOp);
        if (transformErr) return this._hardRollback(transformErr);
      }

      try {
        console.log(this.data);
        console.log(fixupOp.op);
        this._otApply(fixupOp, false);
      } catch (error) {
        return this._hardRollback(error);
      }
    }
  }
```

Then I get as you'd expect: insert in an empty array at index `0`:

```js
{ name: 'fido', tricks: [] }
[ { p: [ 'tricks', 0 ], li: 'stay' } ]
```